aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -347,18 +348,40 @@ impl VMChangeSet {
 
     /// Deposit amount is inserted into metadata at a different time than the WriteOp is created.
     /// So this method is needed to be able to update metadata generically across different variants.
-    pub fn write_set_iter_mut(
-        &mut self,
-    ) -> impl Iterator<Item = (&StateKey, WriteOpSize, &mut StateValueMetadata)> {
-        self.resource_write_set
-            .iter_mut()
-            .map(|(k, v)| (k, v.materialized_size(), v.get_metadata_mut()))
-            .chain(
-                self.module_write_set
-                    .iter_mut()
-                    .chain(self.aggregator_v1_write_set.iter_mut())
-                    .map(|(k, v)| (k, WriteOpSize::from(v as &WriteOp), v.get_metadata_mut())),
-            )
+    pub fn write_op_info_iter_mut<'a>(
+        &'a mut self,
+        executor_view: &'a dyn ExecutorView,
+    ) -> impl Iterator<Item = PartialVMResult<WriteOpInfo>> {
+        let resources = self.resource_write_set.iter_mut().map(|(key, op)| {
+            Ok(WriteOpInfo {
+                key,
+                op_size: op.materialized_size(),
+                prev_size: executor_view
```

Comment:
I am curious: we seem to read here again but I have just realised that using when VM does the write, it reads already, so the size is sort of available thanks to read-before-write property. Wondering if can shift to APIs such that:
- VM can read value
- VM can check existence
- VM knows sizes on read

The whole current design is based on the fact that we have that stupid `ResourceResolver` which fetches bytes, but all of this gas charging can actually go inside the VM, in theory.

cc @igor-aptos @gelash @vgao1996 for thoughts on APIs

Reply:
As we discussed offline, I don't think it would be a good idea to shift this gas charging into the Move VM, nor to transition to a charge-on-the-fly model. The reasoning is that a write is only finalized near the end of a transaction, and the Move VM does not have context on this.

It's possible for the Move VM to annotate the writes in the write set with their original sizes to save a lookup, but I don't think it's worth the trouble for now, assuming the storage is cached.